File: python/146_LRUCache.py
from unittest import TestCase
# https://leetcode.com/problems/lru-cache/


class Node:
    def __init__(self, key=None, value=None, prev=None, next=None):
        self.key = key
        self.value = value
        self.prev = prev
        self.next = next

    def remove(self):
        if self.prev:
            self.prev.next = self.next
        if self.next:
            self.next.prev = self.prev
        self.prev = self.next = None
        return self

    def insert(self, node):
        node.next = self.next
        node.prev = self
        if self.next:
            self.next.prev = node
        self.next = node
        return node


class LRUCache:
    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.count = 0
        self.hash = {}
        self.head = Node()
        self.tail = Node()
        self.head.insert(self.tail)

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        node = self.hash.get(key)
        if node:
            self.move_to_head(node)
            return node.value
        else:
            return -1

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        node = self.hash.get(key)
        if node:
            node.value = value
        else:
            self.hash[key] = self.head.insert(Node(key, value))
            self.count += 1
            if self.count > self.capacity:
                self.pop()
        self.get(key)

    def last(self):
        return self.tail.prev if self.count > 0 else None

    def pop(self):
        if self.count > 0:
            node = self.last()
            del self.hash[node.key]
            node.remove()
            self.count -= 1
            return node
        else:
            return None

    def move_to_head(self, node):
        if self.head.next != node:
            node.remove()
            self.head.insert(node)
        return node


# test:
def to_list(cache):
    list = []
    node = cache.head.next
    while node:
        list.append(node.value)
        node = node.next
    return list[:-1]    # remove tail


class RunTest(TestCase):
    def test1(self):
        c = LRUCache(2)
        c.put(1, 100)
        c.put(2, 200)
        self.assertEqual([200, 100], to_list(c))
        self.assertEqual(100, c.get(1))
        self.assertEqual([100, 200], to_list(c))
        c.put(3, 300)
        self.assertEqual([300, 100], to_list(c))
        c.put(3, 200)
        self.assertEqual([200, 100], to_list(c))
        c.put(3, 300)
        self.assertEqual([300, 100], to_list(c))
        c.put(4, 400)
        self.assertEqual([400, 300], to_list(c))
        self.assertEqual(-1, c.get(1))
        self.assertEqual(300, c.get(3))
        self.assertEqual(400, c.get(4))

File: kotlin/src/main/kotlin/146LRUCache.kt
import org.junit.Assert.assertEquals
import org.junit.Test

//https://leetcode.com/problems/lru-cache/

class Node(
    var key: Int = 0,
    var value: Int = 0,
    var prev: Node? = null,
    var next: Node? = null
){
  fun remove(): Node{
    prev?.next = next
    next?.prev = prev
    next = null
    prev = null
    return this
  }

  fun insert(node: Node): Node{
    node.next = next
    node.prev = this
    next?.prev = node
    next = node
    return node
  }
}

class LRUCache(val capacity: Int){
  var count = 0
  val hash: HashMap<Int, Node> = HashMap()
  val head = Node()
  val tail = head.insert(Node())

  fun get(key: Int): Int {
    if(!hash.containsKey(key)) return -1

    val node = hash[key]!!
    shiftUp(node)
    return node.value
  }

  fun put(key: Int, value: Int): Int{
    if(hash.containsKey(key))
      hash[key]!!.value = value
    else{
      val node = Node(key, value)
      if(count == capacity) pop()
      head.insert(node)
      hash[key] = node
      count ++
    }

    return get(key)
  }

  private fun pop(): Node?{
    if(count == 0) return null

    val node = tail.prev!!.remove()
    hash.remove(node.key)
    count --
    return node
  }

  private fun shiftUp(node: Node): Node{
    if(head.next != node){
      node.remove()
      head.insert(node)
    }
    return node
  }
}

class Tests{
  fun toList(cache: LRUCache): List<Int>{
    val list = mutableListOf<Int>()
    var node = cache.head.next
    while(node != null){
      list.add(node.value)
      node = node.next
    }
    return list.subList(0, list.lastIndex)
  }

  @Test
  fun test(){
    val cache = LRUCache(2)
    cache.put(1, 100)
    cache.put(2, 200)
    assertEquals(listOf(200, 100), toList(cache))
    assertEquals(100, cache.get(1))
    assertEquals(listOf(100, 200), toList(cache))

    assertEquals(100, cache.get(1))
    cache.put(3, 333)
    assertEquals(listOf(333, 100), toList(cache))
    cache.put(3, 300)
    assertEquals(listOf(300, 100), toList(cache))
    cache.put(4, 400)
    assertEquals(-1, cache.get(1))
    assertEquals(400, cache.get(4))
    assertEquals(300, cache.get(3))
  }
}